Adding fzf ebuild.

rafspiny opened the pull request:
Adding ebuild for a new package.

Gentoo Bug: [703082](https://bugs.gentoo.org/703082)
There is one warning that I tried to address, but was not able to. More info on this post.
https://forums.gentoo.org/viewtopic-t-1110938-highlight-.html

juippis commented on app-shells/fzf/fzf-0.21.0.ebuild:
```diff
@@ -0,0 +1,124 @@
+
+LICENSE="MIT"
+SLOT="0"
+KEYWORDS="~amd64 ~arm ~x86"
```
You can only add ~amd64 and x86, provided you tested your ebuild with both arches. Rest **need** to be keyworded through respected arch teams.

juippis commented on app-shells/fzf/fzf-0.21.0.ebuild:
```diff
@@ -0,0 +1,124 @@
+LICENSE="MIT"
+SLOT="0"
+KEYWORDS="~amd64 ~arm ~x86"
+IUSE="fish-completion neovim tmux vim zsh-completion"
```
Add these as `optfeature` and always installed their completion files.
https://projects.gentoo.org/qa/policy-guide/installed-files.html#pg0301

rafspiny replied:
@juippis do you refer to `*-completion` or all of them?
They all install small files. So I guess the second option.

juippis commented on app-shells/fzf/fzf-0.21.0.ebuild:
```diff
@@ -0,0 +1,124 @@
+KEYWORDS="~amd64 ~arm ~x86"
+IUSE="fish-completion neovim tmux vim zsh-completion"
+
+DEPEND=""
+RDEPEND="
+	${DEPEND}
+	fish-completion? ( app-shells/fish )
+	tmux? ( app-misc/tmux )
+	neovim? ( app-editors/neovim )
+	vim? ( app-editors/vim )
+	zsh-completion? ( app-shells/zsh )
+"
+BDEPEND=""
```
All can most likely be removed. Use `optfeature` instead.

juippis commented on app-shells/fzf/fzf-0.21.0.ebuild:
```diff
@@ -0,0 +1,124 @@
+
+EAPI=7
+
+inherit go-module bash-completion-r1
```
You can use an eclass for `zsh` too.

rafspiny replied:
I can't fins zsh eclass listed here: 
https://devmanual.gentoo.org/eclass-reference/

juippis commented on app-shells/fzf/fzf-0.21.0.ebuild:
```diff
@@ -0,0 +1,124 @@
+
+	# Install bash completion file and key-bindings
+	newbashcomp shell/completion.bash ${PN}
+#	mkdir -p "${ED}$(get_bashcompdir)/"
+#	cp shell/completion.bash "${ED}$(get_bashcompdir)/${PN}"
```
Debugging lines should be removed when it works.

rafspiny replied:
Removed

juippis commented on app-shells/fzf/metadata.xml:
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE pkgmetadata SYSTEM "http://www.gentoo.org/dtd/metadata.dtd">
+<pkgmetadata>
+	<maintainer type="person">
+		<email>mgorny@gentoo.org</email>
+		<name>Michał Górny</name>
+	</maintainer>
+	<maintainer type="project">
+		<email>shell-tools@gentoo.org</email>
+		<name>Gentoo Shell Tools Project</name>
```
Are you just adding random people in here hoping they'd start to maintain it? 
You need to become the maintainer unless you have their consent on being added here. You can commit and maintain this package through proxy-maint project, 
https://wiki.gentoo.org/wiki/Project:Proxy_Maintainers/User_Guide#Proxied_maintainer_in_metadata.xml

rafspiny replied:
Copy and paste error

juippis commented:
I also failed to build it, 
```
# github.com/junegunn/fzf
flag provided but not defined: -Wl,--as-needed
usage: link [options] main.o
make: *** [Makefile:103: target/fzf-linux_amd64] Error 2
 * ERROR: app-shells/fzf-0.21.0::testworld failed (compile phase):
 *   emake failed
```

Then you need to read https://www.gentoo.org/glep/glep-0066.html about good commit summaries, and https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin which is required for any commit to be merged into ::gentoo.

juippis commented on app-shells/fzf/fzf-0.21.0.ebuild:
```diff
@@ -0,0 +1,124 @@
+SRC_URI="https://${EGO_PN}/archive/${PV}.tar.gz -> ${P}.tar.gz
+			${EGO_SUM_SRC_URI}"
+
+LICENSE="MIT"
```
And one more thing: You need to include licenses for ALL bundled modules. I think there's an ebuild to check these, go-license-detector or something.

rafspiny commented:
umh. That compilation error is new to me. 
I will double-check LDFLAGS

Alessandro-Barbieri commented:
See also #15407

heroin-moose commented:
@rafspiny, the compilation error is due to missing quotes around -extldflags. See https://github.com/junegunn/fzf/pull/1991.

rafspiny commented:
I had time to look into it today. Fixed it and pushed.
@heroin-moose I arrived to the same conclusion. I created an issue on fzf too and submitted a PR. I looked for LDFLAGS in their issues, but not popped up : /
Oh well, it was not a lot of work in the end.

Everything shall be good now.

juippis commented:
Actually nope, you need to sign-off your commits. And please squash them into one. Describe the package in few words in git message box.

rafspiny commented:
@juippis Sure thing. I will do it tonight

heroin-moose commented:
Is setting S is required? I believe not.

heroin-moose commented:
@juippis, also fzf-tmux will fail if tmux is not installed. Perhaps it makes sense to depend on tmux?

rafspiny commented:
@juippis I rebased and added the `Signed-off-by` tag.
@heroin-moose yep, that secondary binary `fzf-tmux` will fail without tmux. The main binary, `fzf` will still work though.
Do you suggest to put **tmux** as a RDEPEND use flag?

heroin-moose commented:
Yeah, like this:

```
RDEPEND="tmux? ( app-misc/tmux )"
```

rafspiny commented:
@heroin-moose yep. The if part was in my initial commit when `tmux` was still in the `DEPEND` variable.

I am not sure if we shall use the USE flag or not. In the end, that script is supposed to be called from the shell. It is not integrated into other commands. 
On the other hand, I very much dislike having a script on my system that I know is broken.
Probably, the best thing is to re-add the `tmux` use flag.

@juippis what's your suggestions here?

heroin-moose commented:
Yeah, I also don't the like idea of broken scripts in /usr/bin. Along with the man page that states "hey, run this thing, it's good".

heroin-moose commented:
Well... it really works and ignores so maybe there is really no need in USE flag after all.

rafspiny commented:
It does indeed. Even better!

heroin-moose commented:
One more thing: should we install keybindings to /etc for bas? It will overwrite the default ones, which I'm sure that most people want. Maybe it would be better to put all keybindings to /usr/share/fzf.

rafspiny commented:
@heroin-moose I am not sure about that. 
I am doing:
> insinto /etc/bash/bashrc.d/
> newins shell/key-bindings.bash ${PN}.sh

That shall install the file in /etc/bash/bashrc.d/fzf.sh
I am not really following what file is being overwritten

heroin-moose commented:
I mean that built-in bash search function will be replaced with fzf. I'm not sure everyone want that.

rafspiny commented:
@heroin-moose I would argue that replacing the built-in bash search function is the expected behaviour when installing fzf. I would expect to see fzf when I hit CTRL-R

Conflicts with other packages that overwrite the built-in bash search function are another matter

heroin-moose commented:
I, for one, use it as vim plugin and not for shell autocompletion. So I would rather not.

juippis commented:
This package is becoming such a nightmare already to handle... \***sigh**\*

Does it do the same for zsh and fish? I'm thinking at least bash-completion should be behind USE flag then, and the USE flag must describe this behaviour in metadata.xml.

heroin-moose commented:
@juippis, I suggest following Arch way. Installing completions is okay. Keybindings, on the other hand, go to /usr/share/fzf where people can source them if they want.

rafspiny commented:
I use it for both bash and vim. In my original commit, they were under a USE flag.
Debian is also adhering to what you said @heroin-moose 
```
Bash
====

Append this line to ~/.bashrc to enable fzf keybindings for Bash:

   source /usr/share/doc/fzf/examples/key-bindings.bash

Zsh
===

Append this line to ~/.zshrc to enable fzf keybindings for Zsh:

   source /usr/share/doc/fzf/examples/key-bindings.zsh

Fish
====

Issue the following commands to enable fzf keybindings for Fish:

   mkdir -p ~/.config/fish/functions/
   ln -s /usr/share/doc/fzf/examples/fish_user_key_bindings.fish ~/.config/fish/functions/

Vim
===

The most straightforward way is to append the following line to your vimrc:

   source /usr/share/doc/fzf/examples/fzf.vim

```
Shall I do the same and add an `einfo` print. I would do the same for fish and zsh now.
